prelude/src/stdlib/other/IO.sk

```diff
@@ -34,68 +34,108 @@ class OpenOptions{
   }
 }
 
-// TODO: How to handle String vs Bytes? type parameter?
-trait Readable {
-  mutable fun read(sz: Int): String;
-  mutable fun read_to_end(): String;
+// TODO: Proper subclasses mapping to error codes.
+class Error(code: Int) extends Exception uses Show {
+  fun toString(): String {
+    `Error ${this.code}`
+  }
+}
+
+trait Read {
+  mutable fun read(buf: mutable MutableBytes): Result<Int, Error>;
+
+  mutable fun read_to_end(buf: mutable Vector<UInt8>): Result<Int, Error> {
+    nbytes = 0;
+    loop {
+      if (buf.size() == buf.capacity()) {
+        buf.ensureCapacity(buf.capacity() + 32)
```

Comment:
The `ensureCapacity()` method will figure it out and do the right thing:
```
 192 │ // Grow the Vector's internal storage as necessary to accommodate the                                                                                                                             
 193 │ // given (absolute) capacity. This method *may* allocate more storage                                                                                                                             
 194 │ // than requested: if the method succeeds the Vector's capacity will                                                                                                                              
 195 │ // be equal or greater than the given capacity.                                                                                                                                                   
 196 │ mutable fun ensureCapacity(capacity: Int): void {                                                                                                                                                 
 197 │ │ if (capacity > this.capacity()) {                                                                                                                                                               
 198 │ │ │ // Allocate new memory. If the new capacity is close enough to exponential                                                                                                                    
 199 │ │ │ // growth, grow to it. Otherwise use the normal growth heuristic to                                                                                                                           
 200 │ │ │ // avoid O(n^2) behavior (e.g. from calling ensureCapacity(n+1) in a loop).                                                                                                                   
 201 │ │ │ if (capacity < (this.sz * 5).ushr(2)) {                                                                                                                                                       
 202 │ │ │ │ !capacity = getCapacityForSize(capacity)                                                                                                                                                    
 203 │ │ │ };                                                                                                                                                                                            
 204 │ │ │ this.unsafeGrowCapacity(capacity)                                                                                                                                                             
 205 │ │ } else {
```